creusot-contracts/src/resource/fmap_view.rs

```diff
@@ -0,0 +1,195 @@
+    type ViewTy = (K, V);
+    #[logic]
+    #[open]
+    fn view(self) -> (K, V) {
+        self.frag()
+    }
+}
+
+impl<K, V> Authority<K, V> {
+    /// Id of the underlying [`Resource`].
+    #[logic]
+    #[open(self)]
+    pub fn id(self) -> Id {
+        self.0.id()
+    }
+
+    /// Get the authoritative version of the map.
+    ///
+    /// This can also be accessed using the [`view`](crate::View::view) operator `@`.
+    #[logic]
+    #[open(self)]
+    pub fn auth(self) -> FMap<K, V> {
+        match self.0.val().auth {
+            None => FMap::empty(),
+            Some(Excl::Bot) => FMap::empty(),
+            Some(Excl::Excl(auth)) => auth,
+        }
+    }
+
+    /// Create a new, empty authoritative map.
+    #[ensures(result@ == FMap::empty())]
+    pub fn new() -> Ghost<Self> {
+        let resource = Resource::alloc(snapshot!(View::mkauth(FMap::empty())));
+        ghost!(Self(resource.into_inner()))
+    }
+
+    /// Insert a new element in the authoritative map and return the corresponding
+    /// fragment.
+    #[requires(!self@.contains(*k))]
+    #[ensures((^self)@ == self@.insert(*k, *v))]
+    #[ensures((^self).id() == self.id())]
+    #[ensures(result@ == (*k, *v))]
+    #[ensures(result.id() == self.id())]
+    #[pure]
+    #[allow(unused_variables)]
+    pub fn insert(&mut self, k: Snapshot<K>, v: Snapshot<V>) -> Fragment<K, V> {
+        let new_auth = snapshot!(View::mkauth(self.auth().insert(*k, *v)));
+        let new_total = snapshot!(match self.0@.frag {
+            None => *new_auth,
+            Some(frag) => new_auth.op(View::mkfrag(frag)),
+        });
+        self.0.update(new_total);
+        let right = snapshot!(self.0@);
+        let left = snapshot!(View::mkfrag(FMap::empty().insert(*k, Ag::Ag(*v))));
+        Fragment(self.0.split_off(right, left))
+    }
+
+    /// Asserts that the fragment represented by `frag` is contained in `self`.
+    #[requires(self.id() == frag.id())]
+    #[ensures(self.auth().get(frag@.0) == Some(frag@.1))]
+    #[pure]
+    #[allow(unused_variables)]
+    pub fn contains(&self, frag: &Fragment<K, V>) {
+        let new_resource = self.0.join_shared(&frag.0);
+        let new_frag = snapshot!(match new_resource@.frag {
+            None => FMap::empty(),
+            Some(map) => map,
+        });
+        let old_frag = snapshot!(match frag.0@.frag {
+            None => FMap::empty(),
+            Some(map) => map,
+        });
+    }
+}
+
+impl<K, V> Fragment<K, V> {
+    /// Id of the underlying [`Resource`].
+    #[logic]
+    #[open(self)]
+    pub fn id(self) -> Id {
+        self.0.id()
+    }
+
+    /// Get the fragment of the map represented by this resource.
+    ///
+    /// This can also be accessed using the [`view`](crate::View::view) operator `@`.
+    #[logic]
+    #[open(self)]
+    pub fn frag(self) -> (K, V) {
+        let frag_agree = match self.0.val().frag {
+            None => FMap::empty(),
+            Some(frag) => frag,
+        };
```

Comment:
We need something like `unwrap_logic` for this kind of code. This is too verbose.